include/tipsy/binary-to-float.h

```diff
@@ -1,43 +1,123 @@
+#pragma once
+#ifndef TIPSY_ENCODER_BINARY_TO_FLOAT_H
+#define TIPSY_ENCODER_BINARY_TO_FLOAT_H
 /*
  * binary-to-float.h
  *
- * low level encoding to take 3 unsigned chars and project them to a single valid float
- * and vice versa.
+ * Low level encoding to take 3 unsigned chars and project them to a single valid float
+ * and vice versa. The values used are designed for compatability with VCV Rack voltage standards.
  *
  * The encoding we use is as follows:
  *
  * An IEEE float is: lowest 23 bits are fraction, next 8 are exponent, last is sign bit.
  *
  */
-
-#ifndef TIPSY_ENCODER_BINARY_TO_FLOAT_H
-#define TIPSY_ENCODER_BINARY_TO_FLOAT_H
-
 #include <cstdint>
 
 namespace tipsy
 {
-inline float threeBytesToFloat(const unsigned char *const d)
+
+constexpr const unsigned char BYTE_MASK = 0xff;
+constexpr const unsigned char LOW_7_MASK = 0x7f;
+constexpr const unsigned char BIT_8_MASK = 0x80;
+constexpr const unsigned char EXPONENT_FILL = 0x3f;
+
+union FloatBytes
 {
-    unsigned char enc[4];
-    enc[0] = d[0];
-    enc[1] = d[1];
-    enc[2] = d[2] & 127;
-    enc[3] = (d[2] & 128) | 63;
+    unsigned char bytes[4];
+    float f;
+
+    operator float() { return f; }
+
+    unsigned char first() { return bytes[0]; }
+    unsigned char second() { return bytes[1]; }
+    unsigned char third() { return (bytes[2] & LOW_7_MASK) | (bytes[3] & BIT_8_MASK); }
+
+    FloatBytes(float f)
+    {
+        this->f = f;
+    }
+
+    FloatBytes()
+    {
+        bytes[0] = 0;
+        bytes[1] = 0;
+        bytes[2] = 0;
+        bytes[3] = 0x3f;
+    }
+
+    FloatBytes(unsigned char b1, unsigned char b2)
+    {
+        bytes[0] = b1;
+        bytes[1] = b2;
+        bytes[2] = 0;
+        bytes[3] = EXPONENT_FILL;
+    }
+
+    FloatBytes(uint16_t value)
+    {
+        bytes[0] = value & BYTE_MASK;
+        bytes[1] = (value >> 8) & BYTE_MASK;
+        bytes[2] = 0;
+        bytes[3] = EXPONENT_FILL;
+    }
+
+    static bool isRepresentable(uint32_t value)
+    {
+        return 0 == (value & (uint32_t)0xFF000000);
+    }
 
-    float res = *((float *)&(enc[0]));
+    FloatBytes(uint32_t value)
+    {
+        // Not all uint32_t values are representable in this format
+        // Should this state raise an exception? 
+        assert(isRepresentable(value));
+        bytes[0] = value & BYTE_MASK;
+        bytes[1] = (value >> 8) & BYTE_MASK;
+        unsigned char b3 = (value >> 16) & BYTE_MASK;
+        bytes[2] = b3 & LOW_7_MASK;
+        bytes[3] = (b3 & BIT_8_MASK) | EXPONENT_FILL;
+    }
 
-    return res;
+    FloatBytes(unsigned char b1, unsigned char b2, unsigned char b3) {
+        bytes[0] = b1;
+        bytes[1] = b2;
+        bytes[2] = b3 & LOW_7_MASK;
+        bytes[3] = (b3 & BIT_8_MASK) | EXPONENT_FILL;
+    }
+};
+
+inline uint16_t uint16_FromFloat(float f)
+{
+    auto fb = FloatBytes(f);
+    return fb.bytes[0] | (fb.bytes[1] << 8);
```

Comment:
We probably want an assert that byte 3 is zero in this case?

Reply:
ok, I don't think that state is possible, and perhaps it belongs in a test, but not a problem to add it, so I'll make the update.

That brings up a thought I had as I was preparing this PR. What we need is a function that validates that a float fits the schema, but I wasn't sure enough about all the conditions to write it. Some of them are implicit in the binary.cpp tests. I'll open an issue for creating that function.

Reply:
So a simple condition is -10 < f < 10 for fitting into 3 bytes using this encoding 

but the real check is: exponent is the exponent fill. If that is true then it will decode and if it doesn’t then it won’t (since we use all the other bits and fix that fill)